src/bin/customer/pay.rs

```diff
@@ -22,91 +23,137 @@ use super::{connect, database, Command};
-            .send(payment_amount)
-            .await
-            .context("Failed to send payment amount")?
-            .send(note)
-            .await
-            .context("Failed to send payment note")?;
+        let (session_key, chan) = open_session(database.as_ref(), &config, &self.label).await?;
 
-        // Allow the merchant to accept or reject the payment and note
-        offer_abort!(in chan as Customer);
+        let chan = tokio::time::timeout(
+            config.approval_timeout(),
+            request_payment(&config, chan, payment_amount, self.note),
+        )
+        .await
+        .context("Payment timed out while awaiting approval")?
+        .context("Payment was not approved by the merchant")?;
 
         // Run the core zkAbacus.Pay protocol
-        let chan = zkabacus_pay(
-            rng,
-            database.as_ref(),
-            &self.label,
-            session_key,
-            chan,
-            payment_amount,
+        // Timeout is set to 10 messages, which includes all sent & received messages and aborts
+        let chan = tokio::time::timeout(
+            10 * config.message_timeout(),
+            zkabacus_pay(
+                rng,
+                database.as_ref(),
+                &self.label,
+                session_key,
+                chan,
+                payment_amount,
+            ),
         )
         .await
+        .context("Payment timed out while updating channel status")?
         .context("Failed to complete pay protocol")?;
 
-        // Receive the response note (i.e. the fulfillment of the service)
-        let (response_note, chan) = chan
-            .recv()
+        tokio::time::timeout(config.approval_timeout(), receive_service(chan))
             .await
-            .context("Failed to receive response note")?;
+            .context("Payment timed out when receiving service")??;
 
-        // Close the communication channel: we are done communicating with the merchant
-        chan.close();
+        Ok(())
+    }
+}
 
-        // Print the response note on standard out
-        if let Some(response_note) = response_note {
-            eprintln!(
-                "Payment succeeded with response from merchant: \"{}\"",
-                response_note
-            );
-        } else {
-            eprintln!("Payment succeeded with no concluding response from merchant");
-        }
+/// Parse the payment amount specified on the command line.
+fn parse_payment_amount(amount: Amount) -> Result<PaymentAmount, anyhow::Error> {
+    // Convert the payment amount appropriately
+    let minor_units: i64 = amount.try_into_minor_units().ok_or_else(|| {
+        anyhow::anyhow!("Payment amount invalid for currency or out of range for channel")
+    })?;
 
-        Ok(())
+    (if minor_units < 0 {
+        PaymentAmount::pay_customer
+    } else {
+        PaymentAmount::pay_merchant
+    })(minor_units.abs() as u64)
+    .context("Payment amount out of range")
+}
```

Comment:
This seems like it should live in `impl From<Amount> for PaymentAmount`.

Reply:
Yeah, I will update the error message in `PaymentAmount`. This also sneakily maps the error type to `anyhow`, though, so it just gets replaced with a `map_err`.

Reply:
Per offline discussion, moved this entire function into a `TryInto` on `Amount`.